### PR TITLE
Implement vanilla http version

### PR DESCRIPTION
**Pull request by inadarei**

### CODE REVIEW

File: lib/health.js
Comment (randallsquared) [inline lib/health.js]: This doesn't follow the pattern established by the other handlers.   `http` should return the function currently known as `http`, which should be anonymous.  The additional layer for async is still needed, so it's 
```javascript
http() {
  return (request, response) => {
    // L52-65
  };
}
```

And then the usage needs to be 

```javascript
const check = healthcheck();
const handler = check.http();
http.createServer( (request, response) => {
  const isHealthCheckCall = handler(request, response);
  if (isHealthCheckCall) return;
   response.end("HELLO! This is pretty amaziiiiing");
}).listen(3535);
```

**Review by randallsquared — CHANGES_REQUESTED** [review]
I think http() needs work.

File: lib/health.js
Comment (inadarei) [reply lib/health.js]: So, I agree that consistency would be nice, and I like where you are heading but the specific solution you propose won't work since implementation of other functions return an async function and your returns regular one. Koa and Express can take async function as their middleware, Node's vanilla http has no concept of middleware and if we try doing `const isHealthCheckCall = await handler(request, response);` than its surrounding context will have to become async and that can make usage of Maikai really awkward.

Which is why I decided to sacrifice consistency for the sake of "it should be idiomatic for the use-case in which it is used". For somebody using this with vanilla node http - do they care about
1. whether somebody else in Koa uses it the same way
or 
2. whether their usage is close to what they would be doing without healthcheck middleware?

and the decision in the current design was made in favor of #2. 

Thoughts?

Comment (randallsquared) [reply lib/health.js]: I think we agree that there has to be an inconsistency somewhere, given the contexts.

It seems as though you are making an argument that "returns an async function" vs "returns a sync function" is too much inconsistency, but making a change in the interface from "provide a performer" to "perform" is not too much inconsistency.

If that is a fair assessment of our difference, then my argument would be that we should try to contain as much as possible of the difference inside maikai, rather than exposing it to the calling context.  With the proposed changes, the only difference at the call site is that `await` is optional with the `http` handler (since you can await anything at all, providing it does no harm).

Comment (inadarei) [reply lib/health.js]: Here's a possible compromise, if http() returns a non-async function with the following implementation:

```javascript
  http() {
    return (request, response) => {
      const requestPath = request.url;
      if (requestPath !== this.getHealthUri()) return false;

      const self = this;

      (async function () {
        const responseContent = await self.healthResponse(self.opts);
        const httpCode = (response.status === STATUS_ERR ) ? 503 : 200;
        
        response.writeHead(httpCode, { 'Content-Type': 'application/health+json; charset=utf-8' });
        response.end(JSON.stringify(responseContent));
      })();

      return true;
    };
  }
```

then the calling code does become similar to what you proposed:

```javascript
const http = require('http');
const healthcheck = require('../');

http.createServer( (request, response) => {
  const check = healthcheck();
  const checkHandler = check.http();
  const isHealthCheckCall = checkHandler(request, response);
  if (isHealthCheckCall) return;

  response.end("HELLO! This is pretty amaziiiiing");
}).listen(3535);
```

Comment (inadarei) [reply lib/health.js]: Is that what you meant, to begin with, @randallsquared ?

Comment (randallsquared) [reply lib/health.js]: It was exactly what I meant to begin with, yes!    Perhaps I could have been slightly more explicit.  :)

Comment (inadarei) [reply lib/health.js]: Awesome. That sounds great.

Just pushed new version. Can I get LGTM now, please? :)

File: lib/health.js
Comment (jhnlsn) [inline lib/health.js]: if this was `(async () => {` you wouldn't need to redeclare this to self.

File: lib/health.js
Comment (jhnlsn) [inline lib/health.js:70]: Should there be some sort of warning if you are adding a duplicate check?  Here you are just wiping out a check if it already exists.

Comment (inadarei) [reply lib/health.js]: Touche! Fixed.

File: lib/health.js
Comment (inadarei) [reply lib/health.js]: Touche! Fixed

File: lib/health.js
Comment (inadarei) [reply lib/health.js:70]: You mean - if a check *with the same key* already exists? I assume if you call addCheck() with the same key - then you do indeed want to replace the old check? Not a fair assumption?

File: lib/health.js
Comment (jhnlsn) [inline lib/health.js:77]: You are performing a synchronous loop of each check, is this intentional?  My understanding of this code is that if you have 3 checks, you will execute the first check and wait for the response, exec the second check and wait for the response and then exec the 3rd check and wait for the response then finish.  I think it should execute all checks async `Promise.all` style.

Comment (randallsquared) [reply lib/health.js:77]: I agree with this, though it'll require some rework to ensure that none of them throw.   Right now if the executor throws, the `response` from L75 is not used.

File: lib/health.js
Comment (randallsquared) [reply lib/health.js:70]: `addCheck` should throw instead of overwriting, while (a new) `setCheck` should allow overwriting without complaint?

**Review by randallsquared — APPROVED** [review]

**Comment by randallsquared** [comment]
I'm approving even though I think there are lots of improvements, because those aren't what this PR is about.

File: lib/health.js
Comment (inadarei) [reply lib/health.js:77]: Great point. I will address this in the next PR, since that one is entirely about executors.

Thank you

File: lib/health.js
Comment (inadarei) [reply lib/health.js:70]: ok, will incorporate that in the next PR

**Pull request merged**
